src/librustc_mir/interpret/place.rs

```diff
@@ -853,7 +853,7 @@ where
         // We do NOT compare the types for equality, because well-typed code can
         // actually "transmute" `&mut T` to `&T` in an assignment without a cast.
         assert!(
-            src.layout.details == dest.layout.details,
+            src.layout.layout == dest.layout.layout,
```

Comment:
This is confusing, what is this `layout.layout` business?

Reply:
Ah, the first is actually a field of type `TyLayout` which I named `layout` because that seemed reasonable at the time..

Reply:
This is something only miri does, which is to compare the `&Layout` part of `Ty`(`And`)`Layout` while ignoring the `Ty` part. There's a comment above but I'm not 100% sure it makes sense anymore.

Usually you don't see `layout.layout` because a field of `Layout` is accessed through `Ty`(`And`)`Layout` via `Deref`, but here it's getting the `&Layout` references themselves.

This could also be written as `*src.layout == *dest.layout` but I doubt that would be more readable.

Reply:
Yeah I am pretty sure the comment is still accurate: MIR optimizations will turn `x = &*y` into `x = y` even if `y` is a mutable reference. That's why we compare layout ignoring the type.

Reply:
Ah right, this can only happen post-optimizations, I forgot that.

Reply:
Wait, I'm not sure this comparison makes sense except maybe for non-aggregates?
We'll have another look at it, and the other `.layout.layout`s, maybe there are more specific formulations that still perform the right function.

If these are all in asserts it might make more sense to add an exhaustive list of compatible types post-optimizations instead, just so the layout isn't expected to differ in cases in which it might not (e.g. `[usize; 4]` and `[&T; 4]` have the same `Layout`).

Reply:
@RalfJung @oli-obk What about a function like this:
```rust
fn mir_assign_valid_types(dst: Ty<'tcx>, src: Ty<'tcx>) {
    dst == src || match (&dst.kind, &src.kind) {
        (ty::Ref(_, _, dst_pointee), ty::Ref(_, _, src_pointee)) => dst_pointee == src_pointee,
        _ => false,
    }
}
```

Reply:
Well, you can try this easily -- replace [this assertion](https://github.com/rust-lang/rust/blob/master/src/librustc_mir/interpret/place.rs#L864) and run the Miri test suite, that will show of it covers all the cases we see there.

If you prepare a rustc PR I can also offer to run the Miri test suite on it.

Reply:
I opened [an issue](https://github.com/rust-lang/rust/issues/70405), this is preexisting behaviour and not necessary to adjust in this PR